File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/SplashScreen.java
package org.wycliffeassociates.translationrecorder;

import android.content.Intent;
import android.content.res.AssetManager;
import android.media.MediaMetadataRetriever;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.ProgressBar;

import com.door43.tools.reporting.Logger;

import org.apache.commons.codec.binary.Hex;
import org.apache.commons.codec.digest.DigestUtils;
import org.json.JSONException;
import org.wycliffeassociates.translationrecorder.SettingsPage.Settings;
import org.wycliffeassociates.translationrecorder.database.ProjectDatabaseHelper;
import org.wycliffeassociates.translationrecorder.login.UserActivity;
import org.wycliffeassociates.translationrecorder.permissions.PermissionActivity;
import org.wycliffeassociates.translationrecorder.project.ParseJSON;
import org.wycliffeassociates.translationrecorder.project.ProjectPlugin;
import org.wycliffeassociates.translationrecorder.project.components.Language;
import org.wycliffeassociates.translationrecorder.project.components.User;

import java.io.*;
import java.util.List;

/**
 * Created by sarabiaj on 5/5/2016.
 */
public class SplashScreen extends PermissionActivity {

    private static int SPLASH_TIME_OUT = 3000;
    protected ProgressBar mProgressBar;
    private ProjectDatabaseHelper db;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash);

        mProgressBar = (ProgressBar) findViewById(R.id.progress_bar);
        mProgressBar.setMax(100);
        mProgressBar.setIndeterminate(true);
        mProgressBar.setMinimumHeight(8);

        db = ((TranslationRecorderApp)getApplication()).getDatabase();
    }

    @Override
    protected void onPermissionsAccepted() {
        Thread initDb = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    initializePlugins();
                    initDatabase();
                    int profile = PreferenceManager.getDefaultSharedPreferences(SplashScreen.this).getInt(Settings.KEY_PROFILE, -1);
                    if(profile == -1) {
                        Intent intent = new Intent(SplashScreen.this, UserActivity.class);
                        startActivity(intent);
                        finish();
                    } else {
                        startActivity(new Intent(SplashScreen.this, MainMenu.class));
                        finish();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });
        initDb.start();
    }

    private void initDatabase(){
        ParseJSON parse = new ParseJSON(this);
        try {
            //Book[] books = parse.pullBooks();
            Language[] languages = parse.pullLangNames();
            //db.addBooks(books);
            db.addLanguages(languages);
            System.out.println("Proof: en is " + db.getLanguageName("en"));
        } catch (JSONException e) {
            e.printStackTrace();
        }

        importProfiles();
        deleteDanglingProfiles();
    }

    private void initializePlugins() throws IOException {
        AssetManager am = getAssets();
        String[] assetPlugins = am.list("Plugins/Anthologies");
        File pluginsDir = new File(getCacheDir(), "Plugins");
        if (!pluginsDir.exists()) {
            pluginsDir.mkdirs();
        }
        File anthologiesDir = new File(pluginsDir, "Anthologies");
        if (!anthologiesDir.exists()) {
            anthologiesDir.mkdirs();
        }
        copyPluginsFromAssets(am, pluginsDir, anthologiesDir, assetPlugins);

        String[] plugins = anthologiesDir.list();
        if (plugins != null && plugins.length > 0) {
            for (String s : plugins) {
                File plugin = new File(anthologiesDir, s);
                //if (!plugin.exists()) {
                    importPlugin(am, pluginsDir, anthologiesDir, s);
                //}
            }
        }
    }

    private void importPlugin(AssetManager am, File pluginsDir, File anthologiesDir, String plugin) throws IOException {
        File pluginPath = new File(anthologiesDir, plugin);
        ProjectPlugin projectPlugin = new ProjectPlugin(pluginsDir, pluginPath);
        copyPluginContentFromAssets(am, pluginsDir, "Anthologies", plugin);
        copyPluginContentFromAssets(am, pluginsDir, "Books", projectPlugin.getBooksPath());
        copyPluginContentFromAssets(am, pluginsDir, "Versions", projectPlugin.getVersionsPath());
        copyPluginContentFromAssets(am, pluginsDir, "Jars", projectPlugin.getJarPath());


        //copyPluginContentFromAssets(am, pluginPath, "Chunks", projectPlugin.getChunksPath());
        projectPlugin.importProjectPlugin(pluginsDir, db);
    }

    private void copyPluginsFromAssets(AssetManager am, File pluginsDir, File anthologiesDir, String[] plugins) throws IOException {
        if(plugins != null && plugins.length > 0) {
            for(String plugin : plugins) {
                copyPluginContentFromAssets(am, pluginsDir, "Anthologies", plugin);
                File pluginPath = new File(anthologiesDir, plugin);
                ProjectPlugin projectPlugin = new ProjectPlugin(pluginsDir, pluginPath);
                copyPluginContentFromAssets(am, pluginsDir, "Books", projectPlugin.getBooksPath());
                copyPluginContentFromAssets(am, pluginsDir, "Versions", projectPlugin.getVersionsPath());
                copyPluginContentFromAssets(am, pluginsDir, "Jars", projectPlugin.getJarPath());
            }
        }
    }

    private void copyPluginContentFromAssets(AssetManager am, File outputRoot, String prefix, String pluginName) {
        File outputDir = new File(outputRoot, prefix);
        if (!outputDir.exists()) {
            outputDir.mkdirs();
        }
        try (InputStream is = am.open("Plugins/" + prefix + "/" + pluginName);
             FileOutputStream fos = new FileOutputStream(new File(outputDir, pluginName));
        ) {
            byte[] buf = new byte[1024];
            int len;
            while ((len = is.read(buf)) > 0) {
                fos.write(buf, 0, len);
            }
        } catch (IOException e) {
            Logger.e(this.toString(), "Exception copying " + pluginName + " from assets", e);
        }
    }

    private void importProfiles() {
        File profilesDir = new File(
                Environment.getExternalStorageDirectory(),
                getResources().getString(R.string.folder_name) + "/Profiles/"
        );
        if (!profilesDir.exists()) {
            profilesDir.mkdirs();
        }

        for (File profile: profilesDir.listFiles()) {
            String hash = getHash(profile);
            String mimeType = null;

            try {
                MediaMetadataRetriever mmr = new MediaMetadataRetriever();
                mmr.setDataSource(profile.getAbsolutePath());
                mimeType = mmr.extractMetadata(MediaMetadataRetriever.METADATA_KEY_MIMETYPE);
            } catch (Exception e) {
                Log.i("PROFILE", "File is not a media file");
            }

            if(hash != null && mimeType != null && mimeType.equals("audio/mp4")) {
                db.addUser(new User(profile, hash));
            }
        }
    }

    private void deleteDanglingProfiles() {
        List<User> profiles = db.getAllUsers();
        for (User profile: profiles) {
            File file = profile.getAudio();
            if(!file.exists()) {
                db.deleteUser(profile.getHash());
            }
        }
    }

    private String getHash(File file) {
        try {
            return new String(Hex.encodeHex(DigestUtils.md5(new FileInputStream(file))));
        } catch (FileNotFoundException e) {
            return null;
        } catch (IOException e) {
            return null;
        }
    }
}

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/SplashScreen.kt
package org.wycliffeassociates.translationrecorder

import android.app.Activity
import android.content.Intent
import android.content.res.AssetManager
import android.os.Bundle
import android.preference.PreferenceManager
import android.widget.ProgressBar
import com.door43.login.ProfileActivity
import com.door43.login.TermsOfUseActivity
import com.door43.login.core.Profile
import com.door43.tools.reporting.Logger
import org.json.JSONException
import org.wycliffeassociates.translationrecorder.R.id.progress_bar
import org.wycliffeassociates.translationrecorder.SettingsPage.Settings
import org.wycliffeassociates.translationrecorder.data.repository.LanguageRepository
import org.wycliffeassociates.translationrecorder.persistence.repository.RoomDb
import org.wycliffeassociates.translationrecorder.project.ParseJSON
import org.wycliffeassociates.translationrecorder.project.ProjectPlugin
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * Created by sarabiaj on 5/5/2016.
 */
class SplashScreen : Activity() {

    private lateinit var languageDb: LanguageRepository
    private lateinit var roomDb: RoomDb

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)

        roomDb = RoomDb.getInstance(this)
        languageDb = roomDb.languageRepo

        progress_bar as ProgressBar
        progress_bar.max = 100
        progress_bar.isIndeterminate = true
        progress_bar.minimumHeight = 8
    }

    override fun onResume() {
        super.onResume()
        val profile = PreferenceManager.getDefaultSharedPreferences(this).getString(Settings.KEY_PROFILE, "")
        var termsOfUseAccepted = false
        if (profile.isNotBlank()) {
            termsOfUseAccepted = TermsOfUseActivity.termsAccepted(profile, this)
        }
        if (profile.isNotBlank() || !termsOfUseAccepted) {
            val intent = Intent(this, ProfileActivity::class.java)
            intent.putExtra(Profile.PROFILE_KEY, profile)
            startActivityForResult(intent, 42)
        } else {
            val initDb = Thread(Runnable {
                try {
                    initializePlugins()
                    initDatabase()
                    startActivity(Intent(this@SplashScreen, MainMenu::class.java))
                    finish()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            })
            initDb.start()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        //save the profile in preferences
        if (requestCode == 42) {
            if (resultCode == Activity.RESULT_OK) {
                PreferenceManager.getDefaultSharedPreferences(this).edit().putString(Settings.KEY_PROFILE, data.getStringExtra(Profile.PROFILE_KEY)).commit()
                //user backed out of profile page; empty the profile and finish
            } else if (resultCode == Activity.RESULT_CANCELED) {
                PreferenceManager.getDefaultSharedPreferences(this).edit().putString(Settings.KEY_PROFILE, "").commit()
                finish()
                //user backed out of the TOU, keep the profile with unaccepeted TOU and finish. App will resume to TOU page
            } else if (resultCode == TermsOfUseActivity.RESULT_BACKED_OUT_TOU) {
                PreferenceManager.getDefaultSharedPreferences(this).edit().putString(Settings.KEY_PROFILE, data.getStringExtra(Profile.PROFILE_KEY)).commit()
                finish()
                //user declined TOU, clear profile and return to profile page
            } else if (resultCode == TermsOfUseActivity.RESULT_DECLINED_TOU) {
                PreferenceManager.getDefaultSharedPreferences(this).edit().putString(Settings.KEY_PROFILE, "").commit()
            }
        }
    }

    private fun initDatabase() {
        val parse = ParseJSON(this)
        try {
            //Book[] books = parse.pullBooks();
            val languages = parse.pullLangNames()
            //db.addBooks(books);
            languageDb.insertAll(languages.toList())
        } catch (e: JSONException) {
            e.printStackTrace()
        }

    }

    @Throws(IOException::class)
    private fun initializePlugins() {
        val am = assets
        val assetPlugins = am.list("Plugins/Anthologies")
        val pluginsDir = File(cacheDir, "Plugins")
        if (!pluginsDir.exists()) {
            pluginsDir.mkdirs()
        }
        val anthologiesDir = File(pluginsDir, "Anthologies")
        if (!anthologiesDir.exists()) {
            anthologiesDir.mkdirs()
        }
        copyPluginsFromAssets(am, pluginsDir, anthologiesDir, assetPlugins)

        val plugins = anthologiesDir.list()
        if (plugins != null && plugins.isNotEmpty()) {
            for (s in plugins) {
                val plugin = File(anthologiesDir, s)
                //if (!plugin.exists()) {
                importPlugin(am, pluginsDir, anthologiesDir, s)
                //}
            }
        }
    }

    @Throws(IOException::class)
    private fun importPlugin(am: AssetManager, pluginsDir: File, anthologiesDir: File, plugin: String) {
        val pluginPath = File(anthologiesDir, plugin)
        val projectPlugin = ProjectPlugin(pluginsDir, pluginPath)
        copyPluginContentFromAssets(am, pluginsDir, "Anthologies", plugin)
        copyPluginContentFromAssets(am, pluginsDir, "Books", projectPlugin.booksPath)
        copyPluginContentFromAssets(am, pluginsDir, "Versions", projectPlugin.versionsPath)
        copyPluginContentFromAssets(am, pluginsDir, "Jars", projectPlugin.jarPath)

        projectPlugin.importProjectPlugin(this, pluginsDir)
    }

    @Throws(IOException::class)
    private fun copyPluginsFromAssets(am: AssetManager, pluginsDir: File, anthologiesDir: File, plugins: Array<String>?) {
        if (plugins != null && plugins.isNotEmpty()) {
            for (plugin in plugins) {
                copyPluginContentFromAssets(am, pluginsDir, "Anthologies", plugin)
                val pluginPath = File(anthologiesDir, plugin)
                val projectPlugin = ProjectPlugin(pluginsDir, pluginPath)
                copyPluginContentFromAssets(am, pluginsDir, "Books", projectPlugin.booksPath)
                copyPluginContentFromAssets(am, pluginsDir, "Versions", projectPlugin.versionsPath)
                copyPluginContentFromAssets(am, pluginsDir, "Jars", projectPlugin.jarPath)
            }
        }
    }

    private fun copyPluginContentFromAssets(am: AssetManager, outputRoot: File, prefix: String, pluginName: String) {
        val outputDir = File(outputRoot, prefix)
        if (!outputDir.exists()) {
            outputDir.mkdirs()
        }
        try {
            am.open("Plugins/$prefix/$pluginName").use { input ->
                FileOutputStream(File(outputDir, pluginName)).use { fos ->
                    input.copyTo(fos)
                }
            }
        } catch (e: IOException) {
            Logger.e(this.toString(), "Exception copying $pluginName from assets", e)
        }

    }
}
